File: 04_hyperspace/solution/life_expectancy_experiment.jl
using FStrings
using Statistics: mean
using StructArrays

include("parse_file.jl")

predict(X, w) = X * w

loss(X,Y, w) = mean((predict(X, w) - Y) .^ 2)

function gradient(X, Y, w)
    dY = (predict(X, w) - Y)./size(X, 1)
    w_grad = 2 * X' * dY
    return w_grad
end

function train(X, Y; iterations, lr)
    w = zeros(size(X, 2))
    for ii = 0:iterations-1
        # println(f"Iteration {ii:4d} => Loss: {loss(X, Y, w):.10f}")
        dw = gradient(X, Y, w)
        w -= lr*dw
    end
    return w
end

sa = read_file("../../data/life-expectancy/life-expectancy.txt")

X = hcat(ones(length(sa)), sa.Pollution, sa.Healthcare, sa.Water)
Y = sa.Life
w = train(X, Y, iterations=1000*1000, lr=1E-4)

println("Weights: ", w)
println("A few predictions:")
pdct = predict(X, w)
for i = 1:length(sa)
     println(f"{sa.Country[i]} X[{i:d}] -> {pdct[i]:.1f} (label: {Y[i]:.1f}) (diff: {Y[i]-pdct[i]:.2f})")
 end

wlsq = X \ Y
plsq = predict(X, wlsq)
for i = 1:length(sa)
     println(f"{sa.Country[i]} X[{i:d}] -> {plsq[i]:.1f} (label: {Y[i]:.1f}) (diff: {plsq[i]-pdct[i]:.2f})")
 end

println("DONE")

File: 04_hyperspace/solution/life_expectancy_experiment.py
# A solution to the "Hands On: Field Statistician" exercise at the end
# of chapter "Hyperspace!"

import numpy as np


def predict(X, w):
    return np.matmul(X, w)


def loss(X, Y, w):
    return np.average((predict(X, w) - Y) ** 2)


def gradient(X, Y, w):
    return 2 * np.matmul(X.T, (predict(X, w) - Y)) / X.shape[0]


def train(X, Y, iterations, lr):
    w = np.zeros((X.shape[1], 1))
    for i in range(iterations):
        print("Iteration %4d => Loss: %.20f" % (i, loss(X, Y, w)))
        w -= gradient(X, Y, w) * lr
    return w


FILENAME = "../../data/life-expectancy/life-expectancy-without-country-names.txt"
x1, x2, x3, y = np.loadtxt(FILENAME, skiprows=1, unpack=True)
X = np.column_stack((np.ones(x1.size), x1, x2, x3))
Y = y.reshape(-1, 1)
w = train(X, Y, iterations=5_000_000, lr=0.0001)

print("\nWeights: %s" % w.T)
print("\nA few predictions:")
for i in range(10):
    print("X[%d] -> %.4f (label: %d)" % (i, predict(X[i], w).item(), Y[i].item()))
